src/extension.ts

```diff
@@ -186,25 +184,80 @@ export async function activate(ctx: vscode.ExtensionContext) {
           break;
 
         case Options.never:
-          extensionConfigurationSet(configureOnOpenKey, false, vscode.ConfigurationTarget.Workspace);
+          extensionConfigurationSet(SettingsKey.ConfigureOnOpen, false, vscode.ConfigurationTarget.Workspace);
           break;
 
         case Options.yes:
           configureOnOpen = true;
           break;
 
         case Options.always:
-          extensionConfigurationSet(configureOnOpenKey, true, vscode.ConfigurationTarget.Workspace);
+          extensionConfigurationSet(SettingsKey.ConfigureOnOpen, true, vscode.ConfigurationTarget.Workspace);
           configureOnOpen = true;
           break;
       }
     }
+  }
 
-    if (configureOnOpen === true) {
-      runFirstTask("reconfigure");
+  const downloadLanguageServer = extensionConfiguration(SettingsKey.DownloadLanguageServer);
+  const server = extensionConfiguration(SettingsKey.LanguageServer);
+  const shouldDownload = async (downloadLanguageServer: boolean | "ask"): Promise<boolean> => {
+    if (typeof downloadLanguageServer === "boolean")
+      return downloadLanguageServer;
+
+    enum Options {
+      yes = "Yes",
+      no = "Not this time",
+      never = "Never"
+    }
+
+    const response = await vscode.window.showInformationMessage(
+      "Should the extension try to download the language server?",
+      ...Object.values(Options)
+    );
+
+    switch (response) {
+      case Options.yes:
+        extensionConfigurationSet(SettingsKey.DownloadLanguageServer, true, vscode.ConfigurationTarget.Global);
+        return true;
+
+      case Options.never:
+        extensionConfigurationSet(SettingsKey.DownloadLanguageServer, false, vscode.ConfigurationTarget.Global);
+        return false;
+
+      case Options.no:
+        extensionConfigurationSet(SettingsKey.DownloadLanguageServer, "ask", vscode.ConfigurationTarget.Global);
+        return false;
     }
+
+    return false;
   }
 
+  let client = await createLanguageServerClient(server, await shouldDownload(downloadLanguageServer), ctx);
+  if (client !== null) {
+    ctx.subscriptions.push(client);
+    client.start();
+
+    getOutputChannel().appendLine("Not enabling the muon linter/formatter because the language server is active.");
+  } else {
+    activateLinters(root, ctx);
+    activateFormatters(ctx);
+  }
+
+  ctx.subscriptions.push(
+    vscode.commands.registerCommand("mesonbuild.restartLanguageServer", async () => {
+      if (client === null) {
+        client = await createLanguageServerClient(server, await shouldDownload(downloadLanguageServer), ctx);
+        if (client !== null) {
+          ctx.subscriptions.push(client);
+          client.start();
```

Comment:
The output line from above about not enabling muon would be good to have here, though I am not sure of the best factoring.

Reply:
Wouldn't we have to stop linting using muon there? And as far as I can see, there's currently no way to do that

Reply:
Hmm good point. Wanna add a TODO on that front for if we ever figure it out?

Reply:
Yes, will do that